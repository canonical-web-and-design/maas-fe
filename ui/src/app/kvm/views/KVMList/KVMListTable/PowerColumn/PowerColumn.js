import React from "react";
import { useSelector } from "react-redux";

import { capitaliseFirst, getPowerIcon } from "app/utils";
import {
  controller as controllerSelectors,
  machine as machineSelectors,
  pod as podSelectors,
} from "app/base/selectors";
import DoubleRow from "app/base/components/DoubleRow";

const PowerColumn = ({ id }) => {
  const pod = useSelector((state) => podSelectors.getById(state, id));
  const host = useSelector((state) => podSelectors.getHost(state, pod));
  const machinesLoading = useSelector(machineSelectors.loading);
  const controllersLoading = useSelector(controllerSelectors.loading);
  const loading = machinesLoading || controllersLoading;

  const iconClass = getPowerIcon(host, loading);

  let powerText = "Unknown";
  if (host && host.power_state) {
    powerText = capitaliseFirst(host.power_state);
  } else if (!host && loading) {
    powerText = "";
  }

  return (
    <DoubleRow
      icon={<i className={iconClass}></i>}
      iconSpace
      primary={
        <span className="u-nudge-right--small" data-test="pod-power">
          {powerText}
        </span>
      }
    />
  );
};

export default PowerColumn;
