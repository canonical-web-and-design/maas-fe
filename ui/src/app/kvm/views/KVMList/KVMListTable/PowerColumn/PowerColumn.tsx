import React from "react";
import { useSelector } from "react-redux";

import { capitaliseFirst, getPowerIcon } from "app/utils";
import {
  controller as controllerSelectors,
  machine as machineSelectors,
  pod as podSelectors,
} from "app/base/selectors";
import { RootState } from "app/base/types";
import DoubleRow from "app/base/components/DoubleRow";

type Props = { id: number };

const PowerColumn = ({ id }: Props): JSX.Element => {
  const pod = useSelector((state: RootState) =>
    podSelectors.getById(state, id)
  );
  const host = useSelector((state: RootState) =>
    podSelectors.getHost(state, pod)
  );
  const machinesLoading = useSelector(machineSelectors.loading);
  const controllersLoading = useSelector(controllerSelectors.loading);
  const loading = machinesLoading || controllersLoading;

  const iconClass = getPowerIcon(host, loading);

  let powerText = "Unknown";
  if (host && "power_state" in host) {
    powerText = capitaliseFirst(host.power_state);
  } else if (!host && loading) {
    powerText = "";
  }

  return (
    <DoubleRow
      icon={<i className={iconClass}></i>}
      iconSpace
      primary={
        <span className="u-nudge-right--small" data-test="pod-power">
          {powerText}
        </span>
      }
    />
  );
};

export default PowerColumn;
