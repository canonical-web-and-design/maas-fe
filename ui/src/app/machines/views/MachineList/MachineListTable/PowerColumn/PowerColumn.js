import { Spinner } from "@canonical/react-components";
import { useDispatch, useSelector } from "react-redux";
import PropTypes from "prop-types";
import React, { useEffect, useState } from "react";

import { getPowerIcon } from "app/utils";
import { actions as machineActions } from "app/store/machine";
import machineSelectors from "app/store/machine/selectors";
import { useToggleMenu } from "app/machines/hooks";
import DoubleRow from "app/base/components/DoubleRow";

export const PowerColumn = ({ onToggleMenu, systemId }) => {
  const dispatch = useDispatch();
  const [updating, setUpdating] = useState(null);
  const machine = useSelector((state) =>
    machineSelectors.getById(state, systemId)
  );
  const toggleMenu = useToggleMenu(onToggleMenu, systemId);

  const iconClass = getPowerIcon(machine);

  const menuLinks = [];

  const hasOnAction = machine.actions.includes("on");
  const hasOffAction = machine.actions.includes("off");
  const powerState = machine.power_state;
  if (hasOnAction && powerState !== "on") {
    menuLinks.push({
      children: (
        <>
          <i className="p-icon--power-on"></i>Turn on
        </>
      ),
      onClick: () => {
        dispatch(machineActions.on(systemId));
        setUpdating(machine.power_state);
      },
    });
  }
  if (hasOffAction && powerState !== "off") {
    menuLinks.push({
      children: (
        <>
          <i className="p-icon--power-off"></i>Turn off
        </>
      ),
      onClick: () => {
        dispatch(machineActions.off(systemId));
        setUpdating(machine.power_state);
      },
    });
  }
  if (powerState !== "unknown") {
    menuLinks.push({
      children: (
        <>
          <span className="p-table-menu__icon-space"></span>
          Check power
        </>
      ),
      onClick: () => {
        dispatch(machineActions.checkPower(systemId));
        // Don't display the spinner when checking power as we can't reliably
        // determine that the event has finished.
      },
    });
  }
  if (!hasOnAction && !hasOffAction && powerState === "unknown") {
    menuLinks.push({
      children: "No power actions available",
      disabled: true,
    });
  }

  useEffect(() => {
    if (
      updating !== null &&
      (machine.power_state === "error" || machine.power_state !== updating)
    ) {
      setUpdating(null);
    }
  }, [updating, machine.power_state]);

  return (
    <DoubleRow
      icon={
        updating === null ? (
          <i title={powerState} className={iconClass}></i>
        ) : (
          <Spinner
            className="u-no-margin u-no-padding--left u-no-padding--right"
            inline
          />
        )
      }
      iconSpace={true}
      menuClassName="p-table-menu--hasIcon"
      menuLinks={onToggleMenu && menuLinks}
      menuTitle="Take action:"
      onToggleMenu={toggleMenu}
      primary={
        <div className="u-upper-case--first u-truncate" data-test="power_state">
          {powerState}
        </div>
      }
      primaryTitle={powerState}
      secondary={
        <div
          className="u-upper-case--first"
          title={machine.power_type}
          data-test="power_type"
        >
          {machine.power_type}
        </div>
      }
      secondaryTitle={machine.power_type}
    />
  );
};

PowerColumn.propTypes = {
  onToggleMenu: PropTypes.func,
  systemId: PropTypes.string.isRequired,
};

export default React.memo(PowerColumn);
