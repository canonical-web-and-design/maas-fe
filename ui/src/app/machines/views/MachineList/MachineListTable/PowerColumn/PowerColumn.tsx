import { memo, useEffect, useState } from "react";

import PropTypes from "prop-types";
import { useDispatch, useSelector } from "react-redux";

import DoubleRow from "app/base/components/DoubleRow";
import PowerIcon from "app/base/components/PowerIcon";
import { useToggleMenu } from "app/machines/hooks";
import { actions as machineActions } from "app/store/machine";
import machineSelectors from "app/store/machine/selectors";
import type { Machine } from "app/store/machine/types";
import { PowerState } from "app/store/machine/types";
import type { RootState } from "app/store/root/types";
import { NodeActions } from "app/store/types/node";

type Props = {
  onToggleMenu?: (systemId: Machine["system_id"], open: boolean) => void;
  systemId: Machine["system_id"];
};

export const PowerColumn = ({
  onToggleMenu,
  systemId,
}: Props): JSX.Element | null => {
  const dispatch = useDispatch();
  const [updating, setUpdating] = useState<PowerState | null>(null);
  const machine = useSelector((state: RootState) =>
    machineSelectors.getById(state, systemId)
  );
  const toggleMenu = useToggleMenu(onToggleMenu, systemId);
  const powerState = machine?.power_state || PowerState.UNKNOWN;

  useEffect(() => {
    if (
      updating !== null &&
      (powerState === PowerState.ERROR || powerState !== updating)
    ) {
      setUpdating(null);
    }
  }, [powerState, updating]);

  if (!machine) {
    return null;
  }

  const menuLinks = [];
  const hasOnAction = machine.actions.includes(NodeActions.ON);
  const hasOffAction = machine.actions.includes(NodeActions.OFF);
  if (hasOnAction && powerState !== PowerState.ON) {
    menuLinks.push({
      children: <PowerIcon powerState={PowerState.ON}>Turn on</PowerIcon>,
      onClick: () => {
        dispatch(machineActions.on(systemId));
        setUpdating(machine.power_state);
      },
    });
  }
  if (hasOffAction && powerState !== PowerState.OFF) {
    menuLinks.push({
      children: <PowerIcon powerState={PowerState.OFF}>Turn off</PowerIcon>,
      onClick: () => {
        dispatch(machineActions.off(systemId));
        setUpdating(machine.power_state);
      },
    });
  }
  if (powerState !== PowerState.UNKNOWN) {
    menuLinks.push({
      children: (
        <>
          <span className="p-table-menu__icon-space"></span>
          Check power
        </>
      ),
      onClick: () => {
        dispatch(machineActions.checkPower(systemId));
        // Don't display the spinner when checking power as we can't reliably
        // determine that the event has finished.
      },
    });
  }
  if (!hasOnAction && !hasOffAction && powerState === PowerState.UNKNOWN) {
    menuLinks.push({
      children: "No power actions available",
      disabled: true,
    });
  }

  return (
    <DoubleRow
      icon={
        <PowerIcon powerState={powerState} showSpinner={updating !== null} />
      }
      iconSpace={true}
      menuClassName="p-table-menu--hasIcon"
      menuLinks={onToggleMenu && menuLinks}
      menuTitle="Take action:"
      onToggleMenu={toggleMenu}
      primary={
        <div className="u-upper-case--first u-truncate" data-test="power_state">
          {powerState}
        </div>
      }
      primaryTitle={powerState}
      secondary={
        <div
          className="u-upper-case--first"
          title={machine.power_type}
          data-test="power_type"
        >
          {machine.power_type}
        </div>
      }
      secondaryTitle={machine.power_type}
    />
  );
};

PowerColumn.propTypes = {
  onToggleMenu: PropTypes.func,
  systemId: PropTypes.string.isRequired,
};

export default memo(PowerColumn);
