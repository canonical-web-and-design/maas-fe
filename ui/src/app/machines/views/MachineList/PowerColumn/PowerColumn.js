import React from "react";
import classNames from "classnames";
import { useSelector } from "react-redux";

import { machine as machineSelectors } from "app/base/selectors";

const PowerColumn = ({ systemId }) => {
  const machine = useSelector(state =>
    machineSelectors.getBySystemId(state, systemId)
  );

  const iconClass = classNames({
    "p-icon--power-on": machine.power_state === "on",
    "p-icon--power-off": machine.power_state === "off",
    "p-icon--power-error": machine.power_state === "error",
    "p-icon--power-unknown": machine.power_state === "unknown"
  });

  return (
    <div className="p-double-row--with-icon">
      <div className="p-double-row__icon">
        <i title={machine.power_state} className={iconClass}></i>
      </div>
      <div className="p-double-row__primary-row u-upper-case--first">
        <span data-test="power_state">{machine.power_state}</span>
      </div>
      <div className="p-double-row__secondary-row u-upper-case--first">
        <span
          className="u-text-overflow"
          title={machine.power_type}
          data-test="power_type"
        >
          {machine.power_type}
        </span>
      </div>
    </div>
  );
};

export default PowerColumn;
