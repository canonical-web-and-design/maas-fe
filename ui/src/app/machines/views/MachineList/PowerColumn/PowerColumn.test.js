import { MemoryRouter } from "react-router-dom";
import { mount } from "enzyme";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import React from "react";

import PowerColumn from "./PowerColumn";

const mockStore = configureStore();

describe("PowerColumn", () => {
  let state;
  beforeEach(() => {
    state = {
      config: {
        items: []
      },
      machine: {
        errors: {},
        loading: false,
        loaded: true,
        items: [
          {
            system_id: "abc123",
            power_state: "on",
            power_type: "virsh"
          }
        ]
      }
    };
  });

  it("renders", () => {
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <PowerColumn systemId="abc123" />
        </MemoryRouter>
      </Provider>
    );

    expect(wrapper.find("PowerColumn")).toMatchSnapshot();
  });

  it("displays the correct power state", () => {
    state.machine.items[0].power_state = "off";
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <PowerColumn systemId="abc123" />
        </MemoryRouter>
      </Provider>
    );

    expect(wrapper.find('[data-test="power_state"]').text()).toEqual("off");
  });

  it("displays the correct power type", () => {
    state.machine.items[0].power_type = "manual";
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <PowerColumn systemId="abc123" />
        </MemoryRouter>
      </Provider>
    );

    expect(wrapper.find('[data-test="power_type"]').text()).toEqual("manual");
  });
});
