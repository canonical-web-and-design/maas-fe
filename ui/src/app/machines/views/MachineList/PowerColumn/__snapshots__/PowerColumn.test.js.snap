// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PowerColumn renders 1`] = `
<PowerColumn
  systemId="abc123"
>
  <DoubleRow
    icon={
      <i
        className="p-icon--power-on"
        title="on"
      />
    }
    iconSpace={true}
    menuClassName={null}
    menuLinks={
      Array [
        Object {
          "children": <React.Fragment>
            Check power
          </React.Fragment>,
          "onClick": [Function],
        },
      ]
    }
    menuTitle="Take action:"
    primary={
      <div
        className="u-upper-case--first"
        data-test="power_state"
      >
        on
      </div>
    }
    secondary={
      <div
        className="u-upper-case--first"
        data-test="power_type"
        title="virsh"
      >
        virsh
      </div>
    }
  >
    <div
      className="p-double-row--with-icon"
    >
      <div
        className="p-double-row__icon"
      >
        <i
          className="p-icon--power-on"
          title="on"
        />
      </div>
      <div
        className="p-double-row__rows-container"
      >
        <div
          className="p-double-row__primary-row"
        >
          <div
            className="p-double-row__primary-row-text u-truncate"
          >
            <div
              className="u-upper-case--first"
              data-test="power_state"
            >
              on
            </div>
          </div>
          <TableMenu
            className={null}
            links={
              Array [
                Object {
                  "children": <React.Fragment>
                    Check power
                  </React.Fragment>,
                  "onClick": [Function],
                },
              ]
            }
            title="Take action:"
          >
            <ContextualMenu
              className="p-table-menu"
              dropdownClassName="u-no-margin--top"
              hasToggleIcon={true}
              links={
                Array [
                  "Take action:",
                  Object {
                    "children": <React.Fragment>
                      Check power
                    </React.Fragment>,
                    "onClick": [Function],
                  },
                ]
              }
              position="center"
              toggleAppearance="base"
              toggleClassName="u-no-margin--bottom p-table-menu__toggle"
            >
              <span
                className="p-table-menu p-contextual-menu--center"
              >
                <Button
                  appearance="base"
                  aria-controls="Uakgb_J5m9g-0JDMbcJqLJ"
                  aria-expanded="false"
                  aria-haspopup="true"
                  className="p-contextual-menu__toggle u-no-margin--bottom p-table-menu__toggle"
                  hasIcon={true}
                  onClick={[Function]}
                >
                  <button
                    aria-controls="Uakgb_J5m9g-0JDMbcJqLJ"
                    aria-expanded="false"
                    aria-haspopup="true"
                    className="p-button--base has-icon p-contextual-menu__toggle u-no-margin--bottom p-table-menu__toggle"
                    onClick={[Function]}
                  >
                    <i
                      className="p-icon--contextual-menu p-contextual-menu__indicator"
                    />
                  </button>
                </Button>
              </span>
            </ContextualMenu>
          </TableMenu>
        </div>
        <div
          className="p-double-row__secondary-row u-truncate"
        >
          <div
            className="u-upper-case--first"
            data-test="power_type"
            title="virsh"
          >
            virsh
          </div>
        </div>
      </div>
    </div>
  </DoubleRow>
</PowerColumn>
`;
